### PR TITLE
PSP-4: Polkadot domain name service [WIP]

### PR DESCRIPTION
**Pull request by hskang9**
TODOs

- Add hashmap of other blockchain account
- Add a case when parity ink! Smart contract needs domain

### CODE REVIEW

File: psp-1.md [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+* **PSP Number:** 1
+* **Authors:** Hyungsuk Kang(hskang9)
+* **Status:** PoC
+* **Created:** [2019-10-26]
+* **Reference Implementation** 
+[EIP-137: Ethereum Domain Name Service](https://eips.ethereum.org/EIPS/eip-137) 
+
+## Summary
+
+As Ethereum name service did, This draft PSP describes the details of the Polkadot Name Service, a proposed protocol and runtime module function that provides flexible resolution of short, human-readable names to service and resource identifiers. This permits users and developers to refer to human-readable and easy to remember names, and permits those names to be updated as necessary when the underlying resource (contract, content-addressed data, etc) changes.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
As the Ethereum name service did, this draft PSP describes the details of the Polkadot Name Service, a proposed protocol and runtime module function that provides flexible resolution of short, human-readable names to service and resource identifiers. This permits users and developers to refer to human-readable and easy to remember names, and permits those names to be updated as necessary when the underlying resource (contract, content-addressed data, etc.) changes.
```

File: psp-1.md [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+- Multiple services under a single name, such as a DApp hosted in IPFS, a Whisper address, and a mail server.
+- Support for DNS record types, allowing blockchain hosting of 'legacy' names. This would permit an Ethereum client such as Mist to resolve the address of a traditional website, or the mail server for an email address, from a blockchain name.
+- DNS gateways, exposing ENS domains via the Domain Name Service, providing easier means for legacy clients to resolve and connect to blockchain services.
+
+The first two use-cases, in particular, can be observed everywhere on the present-day internet under DNS, and we believe them to be fundamental features of a name service that will continue to be useful as the Polkadot develops and matures.
+
+The normative parts of this document does not specify an implementation of the proposed system; its purpose is to document a protocol that different resolver implementations can adhere to in order to facilitate consistent name resolution. An appendix provides sample implementations of resolver contracts and libraries, which should be treated as illustrative examples only.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
The normative parts of this document do not specify an implementation of the proposed system; its purpose is to document a protocol that different resolver implementations can adhere to in order to facilitate consistent name resolution. An appendix provides sample implementations of resolver contracts and libraries, which should be treated as illustrative examples only.
```

File: psp-1.md [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+
+Likewise, this document does not attempt to specify how domains should be registered or updated, or how systems can find the owner responsible for a given domain. Registration is the responsibility of registrars, and is a governance matter that will necessarily vary between top-level domains.
+
+Updating of domain records can also be handled separately from resolution. Some systems, such as IPFS, may require a well defined interface for updating domains, in which event we anticipate the development of a standard for this.
+
+## Specification
+
+## Overview
+
+ENS in Polkadot just need one part:
```
Comment (Swader) [inline psp-1.md]:
```suggestion
ENS in Polkadot just needs one part:
```

File: psp-1.md [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+- resolver
+
+The resolver is the state where it manages all domains with address. Think of it as a Ethereum world state just for the DNS system.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
The resolver is the state where it manages all domains with addresses. Think of it as an Ethereum world state just for the DNS system.
```

File: psp-1.md [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+    highest_bid: Balance,
+    /// The highest bidder
+    bidder: AccountId,
+    /// Auction closing date in timestamp
+    auction_closed: Moment
+}
+```
+
+The resolver records the auction status and domain information for each item.
+If someone registers a domain which did not exist before, they get domain paying the initilization price(1 milli dot) and have it for 1 year.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
If someone registers a domain which did not exist before, they get the domain by paying the initialization price (1 milli dot) and have it for 1 year.
```

File: psp-1.md [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+The format of the domain(e.g. .io, .com, .dot, etc) can be limited using regular expression in frontend library or in the module using rust `regex` lib in `no_std` mode.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
The format of the domain (e.g. .io, .com, .dot, etc.) can be limited using regular expressions in a frontend library or in the module using rust `regex` lib in `no_std` mode.
```

File: psp-1.md [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+// namehash runs regex if the domain passes the regex, hash with blake2b. Otherwise, throw err. 
+var node = pns.namehash("Bob.dot");
+pns.register(node).then((event) => {
+    if event.status == "registered" {
+        console.log("the name \"Bob.dot\" has been registered by" + account.address);
+    }
+})
+```
+
+the owner can renew the domain before it expires with the price he or she bought.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
The owner can renew the domain before it expires with the price they bought it at. 
```

File: psp-1.md [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+pns.register(node).then((event) => {
+    if event.status == "registered" {
+        console.log("the name \"Bob.dot\" has been registered by" + account.address);
+    }
+})
+```
+
+the owner can renew the domain before it expires with the price he or she bought. 
+
+The auction starts when anyone who want the domain claim auction after ttl from registered date passes or when the owner wants to sell it.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
The auction starts when anyone who wants the domain claims an auction after `ttl` from registered date passes or when the owner wants to sell it.
```

File: psp-1.md [hunk 9]
```diff
@@ -0,0 +1,459 @@
...
+
+The resolver will record the highest bidder only.
+
+Bidding is available before the auction is closed.
+
+The auction will finalize when anyone submits a request to finalize auction after the auction closing date.
+
+The recorded highest bidder will take the domain and register it new with the new price and have it for one year.
```
Comment (Swader) [inline psp-1.md]:
```suggestion
The recorded highest bidder will take the domain and register it with the new price and have it for one year. 
```

File: psp-1.md [hunk 10]
```diff
@@ -0,0 +1,459 @@
...
+fn resolve(origin, domain_hash: T::Hash) -> Result
+```
+
+resolves the domain and emits event `DomainResolved`
+
+```
+fn renew(origin, domain_hash: T::Hash) -> Result
+```
+
+renews the domain and emits event `DomainRenewal`
```
Comment (Swader) [inline psp-1.md]:
```suggestion
renews the domain and emits event `DomainRenewed`
```

Keeping the tense in the past for consistency with above events.

File: psp-1.md [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+			let ttl = Self::to_milli(T::Moment::from(YEAR));
+			new_domain.ttl += ttl;		
+
+			// Try to withdraw price from the user account to renew the domain 
+			let _ = <balances::Module<T> as Currency<_>>::withdraw(&sender, new_domain.price, WithdrawReason::Reserve, ExistenceRequirement::KeepAlive)?;			
+
+
+			// mutate domain with new_domain struct in the Domain state
+			<Resolver<T>>::mutate(domain_hash.clone(), |domain| *domain = new_domain.clone());
+			Self::deposit_event(RawEvent::DomainRenewal(domain_hash, sender, new_domain.registered_date + new_domain.ttl));
```
Comment (Swader) [inline psp-1.md]:
```suggestion
			Self::deposit_event(RawEvent::DomainRenewed(domain_hash, sender, new_domain.registered_date + new_domain.ttl));
```

File: psp-1.md [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+
+decl_event!(
+	pub enum Event<T> where AccountId = <T as system::Trait>::AccountId, <T as system::Trait>::Hash, <T as balances::Trait>::Balance, <T as timestamp::Trait>::Moment
+ {
+		DomainRegistered(AccountId, Balance, Moment, Moment),
+		NewAuction(AccountId, Hash, Moment, Moment), 
+		NewBid(AccountId, Hash, Balance),
+		AuctionFinalized(AccountId, Hash, Balance),
+		DomainResolved(Hash, AccountId),
+		DomainRenewal(Hash, AccountId, Moment),
```
Comment (Swader) [inline psp-1.md]:
```suggestion
		DomainRenewed(Hash, AccountId, Moment),
```

Comment (hskang9) [reply psp-1.md]: Thank you for your review. It was very interesting to review them.

1) I was thinking to make a hashmap in the Domain struct and use key as the chain index from [slips-44](https://github.com/satoshilabs/slips/blob/master/slip-0044.md), value as the byte array(Vec<u8>) of the corresponding blockchain address. I see some missing address schemes, so I might come up with the new index for polkadot extending the existing chain index. If you have other identifiers, please let me know.

2) I saw ENS has been working on [javascript client side library](https://github.com/ensdomains/address-encoder) to provide encoding/decoding service for the address in multi chains. I think I will make the library compatible with this standard. In fact, I think someone can help ENS's library adding polkadot address scheme as a [bounty](https://explorer.bounties.network/bounty/3584). The only difference would be the address string will be converted to a byte array(Vec<u8>) to store in Substrate. I think I should refer this too in the proposal.

3) Reverse lookup is possible, and I think I will use this data to check referral to encompass new users trying to use this service. Naming service requires a cryptocurrency to pay fee for interaction between blockchain. ENS requires ether, and the user must get one from crypto exchanges or his or her friend. I want to eliminate this barrier using native blockchain to let the general public use this service via referral. 

The process works like this:

1. User A makes a domain from the name service and registers its address to the Reverse lookup storage.

2. User B without any cryptocurrency balance wants to make a domain name, and User A sends the extrinsic that shows the intent that User B wants to register the domain.

3. The blockchain checks the Reverse lookup storage and checks User B exists in the Reverse lookup storage. If User B does not exist, the blockchain adds User B in the reverse lookup storage with a null value and mints 0.001 DOT(new domain registration fee) + extrinsic fees to User B. 

4. User B gets the fund to use domain name service and registers the new one for his or her account.

**Comment by hskang9** [comment]
@Swader I added reverse registry in the runtime. please check including typos.

**Comment by hskang9** [comment]
I was trying to add `claim(byte32 node)` just as ENS, but I found no logic to check if the claimer is the owner of the domain. Eventually, I integrated to the domain transfer logic.

**Comment by Noc2** [comment]
This should be PSP number 4 (=4th PR to the repo) instead of 1.

**Comment by burdges** [comment]
Please get feedback form Wei @sorpaas and @shawntabrizi.  They've some code already and may have opinions on how this overlaps with other things.

**Pull request closed without merging**